field/src/coset.rs

```diff
@@ -0,0 +1,222 @@
+                shift: self.shift,
+                log_size: new_log_size,
+            })
+    }
+
+    /// Returns the coset `self^(2^log_scale_factor)` (i. e. with shift and
+    /// subgroup generator equal to the `2^log_scale_factor`-th power of the
+    /// original ones). Returns `None` if `log_scale_factor` is greater than
+    /// factor is larger than size of `self.log_size()`.
+    pub fn exp_power_of_2(&self, log_scale_factor: usize) -> Option<Self> {
+        self.shrink_coset(log_scale_factor).map(|mut coset| {
+            coset.shift = self.shift.exp_power_of_2(log_scale_factor);
+            coset
+        })
+    }
+
+    /// Returns a new coset where the shift has been set to `shift` times the
+    /// original shift.
+    pub fn shift_by(&self, shift: F) -> TwoAdicMultiplicativeCoset<F> {
+        TwoAdicMultiplicativeCoset {
+            shift: self.shift * shift,
+            log_size: self.log_size,
+        }
+    }
+
+    /// Returns a new coset where the shift has been set to `shift`
+    pub fn set_shift(&self, shift: F) -> TwoAdicMultiplicativeCoset<F> {
+        TwoAdicMultiplicativeCoset {
+            shift,
+            log_size: self.log_size,
+        }
+    }
+
+    /// Checks if the given field element is in the coset
+    pub fn contains(&self, element: F) -> bool {
+        // Note that, in a finite field F (this is not true of a general finite
+        // commutative ring), there is exactly one subgroup of |F^*| of order n
+        // for each divisor n of |F| - 1, and its elements e are uniquely
+        // caracterised by the condition e^n = 1.
+
+        // We check (shift^{-1} * element)^(2^log_size) = 1, which is equivalent
+        // to checking shift^(2^log_size) = element^(2^log_size) - this avoids
+        // inversion at the cost of a few squarings. The loop terminates early
+        // if possible.
+        let (mut shift, mut element) = (self.shift, element);
+
+        for _ in 0..self.log_size {
+            if element == shift {
+                return true;
+            }
+            element = element.square();
+            shift = shift.square();
+        }
+
+        element == shift
+    }
+
+    /// Returns the element `shift * generator^index`, which is the `index %
+    /// self.size()`-th element of `self` (and, in particular, the `index`-th
+    /// element of `self` whenever `index` < self.size()).
+    #[inline]
+    pub fn element(&mut self, index: usize) -> F {
+        self.shift * self.generator_exp(index & ((1 << self.log_size) - 1))
+    }
+
+    // Internal function which computes `generator^exp`. It uses the
+    // square-and-multiply algorithm with the caveat that squares of the
+    // generator are queried from the field (which typically should have them
+    // stored), i. e. rather "fetch-and-multiply".
+    fn generator_exp(&mut self, exp: usize) -> F {
+        let mut gen_power = F::ONE;
+        let mut exp = exp & ((1 << self.log_size()) - 1);
```

Comment:
Up to you but thought this could do with a short comment for clarity. Also we can use `self.size()` instead of `(1 << self.log_size()) - 1`. (I realize that `(1 << self.log_size()) - 1` was my suggestion originally, just noticed the improvement now)
```suggestion
        // As `generator` satisfies `generator^{self.size()} == 1` we can replace `exp` by `exp mod self.size()`.
        // As `self.size()` is a power of `2` this can be done with an `&` instead of a `%`.
        let mut exp = exp & (self.size() - 1);
```